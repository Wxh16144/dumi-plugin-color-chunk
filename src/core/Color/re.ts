// <http://www.w3.org/TR/css3-values/#integers>
const CSS_INTEGER = '[-\\+]?\\d+%?';

// <http://www.w3.org/TR/css3-values/#number-value>
const CSS_NUMBER = '[-\\+]?\\d*\\.\\d+%?';

// Allow positive/negative integer/number.  Don't capture the either/or, just the entire outcome.
const CSS_UNIT = `(?:${CSS_NUMBER})|(?:${CSS_INTEGER})`;

// like: https://github.com/scttcper/tinycolor/blob/31a98e09ac0f9499ccf6971acd2a006703527d27/src/format-input.ts#L98-L99
const PERMISSIVE_MATCH3_STRICT = `\\s*\\(\\s*(${CSS_UNIT})\\s*,\\s*(${CSS_UNIT})\\s*,\\s*(${CSS_UNIT})\\s*\\)\\s*$`;
const PERMISSIVE_MATCH4_STRICT = `\\s*\\(\\s*(${CSS_UNIT})\\s*,\\s*(${CSS_UNIT})\\s*,\\s*(${CSS_UNIT})\\s*,\\s*(${CSS_UNIT})\\s*\\)\\s*$`;

/**
 * like: https://github.com/scttcper/tinycolor/blob/31a98e09ac0f9499ccf6971acd2a006703527d27/src/format-input.ts#L101
 */
export const strictMatchers = {
  rgb: new RegExp(`^rgb${PERMISSIVE_MATCH3_STRICT}`),
  rgba: new RegExp(`^rgba${PERMISSIVE_MATCH4_STRICT}`),
  hsl: new RegExp(`^hsl${PERMISSIVE_MATCH3_STRICT}`),
  hsla: new RegExp(`^hsla${PERMISSIVE_MATCH4_STRICT}`),
  hsv: new RegExp(`^hsv${PERMISSIVE_MATCH3_STRICT}`),
  hsva: new RegExp(`^hsva${PERMISSIVE_MATCH4_STRICT}`),
  cmyk: new RegExp(`^cmyk${PERMISSIVE_MATCH4_STRICT}`),
  hex3: /^#([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/,
  hex6: /^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/,
  hex4: /^#([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})([0-9a-fA-F]{1})$/,
  hex8: /^#([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$/,
};
