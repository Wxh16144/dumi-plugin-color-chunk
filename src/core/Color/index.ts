import { Color as TinyColor } from '@rc-component/color-picker';
import { strictMatchers } from './re';

const map = new Map<string, boolean>();

class Color extends TinyColor {
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(color: string = '') {
    super(color.trim());
  }

  /**
   * 严格的校验
   */
  public isStrictValid(): boolean {
    if (this.format === 'name') return true;
    if (typeof this.originalInput !== 'string') return false;

    if (map.has(this.originalInput)) {
      return map.get(this.originalInput) as boolean;
    } else if (this.originalInput) {
      const _r = Object.values(strictMatchers).some((re) => re.test(this.originalInput as string));
      map.set(this.originalInput, _r);
      return _r;
    }

    return false;
  }
}

export default Color;
