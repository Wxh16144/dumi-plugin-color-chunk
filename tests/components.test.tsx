import { describe, expect, it } from 'vitest';
import Color from '../src/core/Color';

describe('export', () => {
  it('should work', async () => {
    // @ts-ignore
    const all: any = await import('dumi-plugin-color-chunk/component');

    expect(Object.keys(all)).toMatchSnapshot();
    expect(all.Color).equal(Color);
  });
});
