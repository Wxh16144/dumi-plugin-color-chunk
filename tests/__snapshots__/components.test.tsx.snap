// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`export > should work 1`] = `
[
  "TinyColor",
  "names",
  "readability",
  "isReadable",
  "mostReadable",
  "toMsFilter",
  "fromRatio",
  "legacyRandom",
  "inputToRGB",
  "stringInputToObject",
  "isValidCSSUnit",
  "random",
  "bounds",
  "rgbToRgb",
  "rgbToHsl",
  "hslToRgb",
  "rgbToHsv",
  "hsvToRgb",
  "rgbToHex",
  "rgbaToHex",
  "rgbaToArgbHex",
  "cmykToRgb",
  "rgbToCmyk",
  "convertDecimalToHex",
  "convertHexToDecimal",
  "parseIntFromHex",
  "numberInputToObject",
  "Color",
  "default",
]
`;
